HACBS Commits List

abhinandan13jan opened the pull request:
## Fixes 
<!-- For e.g Fixes: https://issues.redhat.com/browse/HAC-XXX -->
https://issues.redhat.com/browse/HACBS-824
https://issues.redhat.com/browse/HACBS-819

## Description
<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Adds commits table in application commits tab and recent commits in overview tab. The table is populated with mock data as we don't have sufficient live data as of today.


## Type of change
<!-- Please delete options that are not relevant. -->

- [x] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Screen shots / Gifs for design review 
<!-- If change affects UI in any way, tag relevant UX people and add screenshots/gifs  -->
<img width="1499" alt="Screenshot 2022-08-19 at 5 16 01 PM" src="https://user-images.githubusercontent.com/24852534/185613168-be07a020-1ef8-4053-b530-4235df375dd3.png">
<img width="1792" alt="Screenshot 2022-08-19 at 5 16 12 PM" src="https://user-images.githubusercontent.com/24852534/185613179-5426007e-8bf1-46d1-98ab-ec26651eae2d.png">
<img width="1497" alt="Screenshot 2022-08-19 at 5 16 27 PM" src="https://user-images.githubusercontent.com/24852534/185613185-d34a0c21-6b78-417d-876c-29d4e447cd4c.png">
<img width="1498" alt="Screenshot 2022-08-19 at 5 16 43 PM" src="https://user-images.githubusercontent.com/24852534/185613189-aa3d608b-1a86-44f0-b2e1-3a2822675e57.png">
<img width="1495" alt="Screenshot 2022-08-19 at 5 16 55 PM" src="https://user-images.githubusercontent.com/24852534/185613192-5b5f9553-c7e4-4a6f-9812-c5c496f30a87.png">
<img width="1792" alt="Screenshot 2022-08-19 at 5 17 10 PM" src="https://user-images.githubusercontent.com/24852534/185613195-4ca01470-2abf-4673-8b4c-fa900ea1edcf.png">
<img width="1792" alt="Screenshot 2022-08-17 at 5 25 20 PM" src="https://user-images.githubusercontent.com/24852534/185119562-8e06445a-7138-41a8-a5df-5f1f0d4de64e.png">


## How to test or reproduce?
<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

<!-- - [ ] Test A -->
<!-- - [ ] Test B -->

<!-- **Test Configuration(s)**: -->


## Browser conformance: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge


<!-- ## Checklist: -->

<!-- 
- [ ] Code follows the style guidelines
- [x] Self-reviewed the code
- [ ] Added comments in hard-to-understand areas
- [ ] Made corresponding changes to the documentation
- [ ] Changes generate no new warnings
- [ ] Added tests that prove this fix is effective or that the feature works
- [ ] New and existing unit tests pass locally with new changes
- [ ] Any dependent changes have been merged and published in downstream modules 
-->

Ranelim commented:
@abhinandan13jan Thank you for putting this together, looks good so far!

Some missing things:
- Tab title
- Tab description
- In the name column, adding the Github logo to jump to that commit on github
- Statuses within the status column, and their interactions (I gather those will be added later after we have a commit details page?)
- Search bar, that search by commit or ID name (toggle or direct search)
- Empty state 

See [this image](https://docs.google.com/document/d/1vlTQOmdwxHe6zMMLHjMjQ0Mi66uoR7_gPECbNP4d8T4/edit#bookmark=id.ntrckt1txwc4) in the deliverables doc
Many thanks!

abhinandan13jan commented:
@Ranelim Updated please have a look.

abhinandan13jan commented:
Status part will have to update later when we have a clear util/definition for a commit status

karthikjeeyar commented on src/hacbs/components/Commits/__mocks__/pipeline-with-commits.ts:
```diff
@@ -0,0 +1,725 @@
+        'build.appstudio.redhat.com/user': 'abhi',
+      },
+      name: 'java-springboot-sample-x778q',
+      namespace: 'abhindas',
```
Change it to `test` or `test-ns` all the occurrences
```suggestion
      namespace: 'test-ns',
```

abhinandan13jan replied:
updated

karthikjeeyar commented on src/hacbs/components/Commits/__mocks__/pipeline-with-commits.ts:
```diff
@@ -0,0 +1,725 @@
+import { PipelineRunKind } from '../../../types';
```
`__mocks__` directory is used for [mocking a component or a module in jest](https://jestjs.io/docs/manual-mocks#mocking-user-modules). Let us use `__data__` to place test data.

abhinandan13jan replied:
updated

karthikjeeyar commented on src/hacbs/components/Commits/CommitsListRow.tsx:
```diff
@@ -0,0 +1,49 @@
+      <TableData className={commitsTableColumnClasses.committedAt}>
+        <Timestamp timestamp={obj.creationTime} />
+      </TableData>
+      <TableData className={commitsTableColumnClasses.status}>-</TableData>
```
Status Column should be populated here. There could be multiple pipelineruns associated with a commit, should we show the latest pipelinerun status here?

karthikjeeyar commented on src/hacbs/utils/commits-utils.ts:
```diff
@@ -0,0 +1,77 @@
+import { PipelineRunKind } from '../types';
+
+type Commit = {
```
Move this to `hacbs/types/commit.ts`

karthikjeeyar commented on src/hacbs/components/Commits/AppRecentCommits.tsx:
```diff
@@ -0,0 +1,38 @@
+  const commits = pipelineRunstoCommits(pipelineWithCommits, COMMIT_LIMIT);
+  const navigate = useNavigate();
+  return (
+    <PageSection padding={{ default: 'noPadding' }} variant={PageSectionVariants.light} isFilled>
+      <Title size="lg" headingLevel="h3" className="pf-c-title pf-u-mt-lg pf-u-mb-sm">
+        Recent commits
+      </Title>
+      <Text className="pf-u-mb-lg">
+        Monitor your commits and their pipeline progression across all components
+        <Button variant="link" className="pf-u-pl-sm">
+          learn more
+        </Button>
+      </Text>
+      <CommitsListView commits={commits} />
```
we can avoid this duplication by separating it out to a new abstract component that accepts props for title and action buttons and use it in `commitsTab` and in here?

abhinandan13jan replied:
updated accordingly

karthikjeeyar commented on src/hacbs/components/Commits/CommitsListRow.tsx:
```diff
@@ -0,0 +1,49 @@
+import { Timestamp } from '../../../shared/components/timestamp/Timestamp';
+import { commitsTableColumnClasses } from './CommitsListHeader';
+
+const CommitsListRow: React.FC<RowFunctionArgs<any>> = ({ obj }) => {
```
replace `<any>` with proper type and add type for obj .

karthikjeeyar commented on src/hacbs/utils/commits-utils.ts:
```diff
@@ -0,0 +1,77 @@
+export const COMMIT_USER_LABEL = 'build.appstudio.redhat.com/user';
+export const COMMIT_REPO_URL = 'build.appstudio.redhat.com/sha_url';
+
+export const pipelineRunstoCommits = (plrList: PipelineRunKind[], limit?: number) => {
```
return type is missing

Ranelim commented:
Thank you @abhinandan13jan 
This looks really good!


Can the search bar search the name of the commit and the commit id?
If so, can we reflect that in the search bar placeholder text, and drop the "name" filter?
If not, and that functionality changes with the filter, the dropdown that switches between commit id and commits name should be a dropdown.

Lastly, is there an empty state?

Many thanks

karthikjeeyar commented on src/hacbs/components/Commits/CommitsListView.tsx:
```diff
@@ -0,0 +1,63 @@
+        </ToolbarContent>
+      </Toolbar>
+      <Table
+        data={filteredComits}
```
Filtered empty state contains default `not found` text

![image](https://user-images.githubusercontent.com/9964343/185400264-2835b82c-1def-4aaa-94c6-acc1241fc7bf.png)


When the filtered results are empty, we show an empty state in other places.
here is a screenshot from Christian's Environment PR:  `Settings -> Environment`.  we can follow this as a convention in all the list view with filters. cc:  @christianvogt  @Ranelim

![image](https://user-images.githubusercontent.com/9964343/185399014-5d451378-849c-41c7-94cd-8468eede7405.png)

abhinandan13jan commented:
@karthikjeeyar  I've updated the screenshots as per latest discussion. Merged the two use cases of Recent and regular commits into a single component rendered as per `recentOnly` props. Updated to use a defined empty state rather than default Table empty state.
@Ranelim I've dropped the name dropdown and am using a `SearchInput`(refer to the screenshots) as per patterfly Toolbar guidelines. let me know if we need to update it. The `commitid` is actually the commit Name. The annotation only provides two things `sha_id`(commitid) & `sha_url` (github url for the commit ). Let me know what placeholder value is to be used.

P.S: I've left the Status unpopulated as we dont have an definition/understanding of commit status as of today. I am storing all the PLR status corresponding to a commit. Once we have the status definition as per the commits details story we can apply the same function here.

karthikjeeyar commented on src/hacbs/utils/commits-utils.ts:
```diff
@@ -0,0 +1,62 @@
+        if (commits[existingCommitIndex].creationTime < creationTime) {
+          commits[existingCommitIndex].creationTime = creationTime;
+        }
+        if (!commits[existingCommitIndex].component) {
+          commits[existingCommitIndex].component = component;
+        }
```
As per design [doc](https://docs.google.com/document/d/1vlTQOmdwxHe6zMMLHjMjQ0Mi66uoR7_gPECbNP4d8T4/edit#heading=h.xes6l26uie9e), Components is a comma separated string if multiple components are involved.

karthikjeeyar commented on src/hacbs/utils/commits-utils.ts:
```diff
@@ -0,0 +1,62 @@
+        if (!commits[existingCommitIndex].component) {
+          commits[existingCommitIndex].component = component;
+        }
+        if (!commits[existingCommitIndex].repoURL) {
+          commits[existingCommitIndex].repoURL = repoURL;
+        }
+        if (!commits[existingCommitIndex].shaURL) {
+          commits[existingCommitIndex].shaURL = shaURL;
+        }
+        if (!commits[existingCommitIndex].user) {
+          commits[existingCommitIndex].user = commitUser;
+        }
```
isn't these are some common fields to the pipelineruns that belongs to the same commit, do we really need this to set it on every iteration?

abhinandan13jan replied:
This only sets if this parameter(key) for the commit is non-existent.

karthikjeeyar commented on src/hacbs/utils/commits-utils.ts:
```diff
@@ -0,0 +1,62 @@
+import { PipelineRunKind, Commit } from '../types';
+
+export const COMMIT_ANNOTATION = 'build.appstudio.redhat.com/commit_sha';
+export const COMMIT_URL_ANNOTATION = 'build.appstudio.redhat.com/sha_url';
+export const COMMIT_BRANCH_ANNOTATION = 'build.appstudio.redhat.com/target_branch';
+export const COMMIT_COMPONENT_LABEL = 'build.appstudio.redhat.com/component';
+export const COMMIT_USER_LABEL = 'build.appstudio.redhat.com/user';
+export const COMMIT_REPO_URL = 'build.appstudio.redhat.com/sha_url';
```
move these consts to `PipelineRunLabel` enum in `src/hacbs/consts/pipelinerun.ts` file

karthikjeeyar commented on src/hacbs/components/Commits/commit-actions.ts:
```diff
@@ -0,0 +1,12 @@
+    {
+      cta: () => {},
+      id: `view-${commit.metadata.name.toLowerCase()}`,
+      label: 'View',
```
`Go to source` kebab action is outlined in the design doc. 
<img width="251" alt="image" src="https://user-images.githubusercontent.com/9964343/185913004-339f9f74-91c9-4438-bbaa-b8ee5c340a8e.png">

karthikjeeyar commented on src/hacbs/components/Commits/CommitsListView.tsx:
```diff
@@ -0,0 +1,122 @@
+          learn more
+        </Button>
+      </Text>
+      <Toolbar data-testid="component-list-toolbar" clearAllFilters={onClearFilters}>
```
### 
```suggestion
      <Toolbar data-test="commit-list-toolbar" clearAllFilters={onClearFilters}>
```

karthikjeeyar commented on src/hacbs/components/Commits/CommitsListView.tsx:
```diff
@@ -0,0 +1,122 @@
+        No results match the filter criteria. Remove filters or clear all filters to show results.
+      </EmptyStateBody>
+      <EmptyStateSecondaryActions>
+        <Button variant="link" onClick={onClearFilters} data-test-id="catalog-clear-filters">
```
```suggestion
        <Button variant="link" onClick={onClearFilters} data-test="commit-clear-filters">
```

karthikjeeyar reviewed (commented):
FYI @abhinandan13jan PAC is now enabled in HACBS workflow, we no longer need `commit` mock data to populate the list. Ref [thread](https://coreos.slack.com/archives/C038DJAP7HR/p1660851449707629)

karthikjeeyar commented on src/hacbs/components/Commits/CommitsListView.tsx:
```diff
@@ -0,0 +1,122 @@
+                type="search"
+                aria-label="name filter"
+                placeholder="Filter by name..."
+                onChange={(name) => onNameInput(name)}
```
After clearing the filter, the search term is not cleared. 
```suggestion
                onChange={(name) => onNameInput(name)}
                value={nameFilter}
```

karthikjeeyar commented on src/hacbs/components/ApplicationDetails/tabs/CommitsTab.tsx:
```diff
@@ -9,13 +9,17 @@ import {
 
 const CommitsTab: React.FC<CommitTabProps> = ({ applicationName }) => {
-  return (
+  const commits = getCommitsFromPLRs(pipelineWithCommits);
```
We need to remove the mock data and use actual pipelineruns
```suggestion
  const commits = getCommitsFromPLRs(pipelineWithCommits);
```

Ranelim commented:
@abhinandan13jan Looks excellent! Thank you!

From the overview tab, we don't need the search box for the recent commits. It is 5 tops. 

Thank you for all the excellent work!

karthikjeeyar commented:
Brackets around a value is used as placeholders in the designs,  I think the branch name should be displayed as it is without square brackets `[main]`  ---> `main`. @Ranelim could you please confirm this?

![image](https://user-images.githubusercontent.com/9964343/186330681-dc9f0c38-62a2-4a0e-839e-bee0b803c0b4.png)

karthikjeeyar commented on src/hacbs/components/Commits/AppRecentCommits.tsx:
```diff
@@ -0,0 +1,80 @@
+import { getCommitsFromPLRs } from '../../utils/commits-utils';
+import CommitsListView from './CommitsListView';
+
+const COMMIT_LIMIT = 3;
```
Ran mentioned top 5 commits in his previous comment

> From the overview tab, we don't need the search box for the recent commits. It is 5 tops.

karthikjeeyar commented on src/hacbs/components/Commits/CommitsListRow.tsx:
```diff
@@ -0,0 +1,64 @@
+        )}
+      </TableData>
+      <TableData className={commitsTableColumnClasses.branch}>
+        {obj.gitProvider === 'github' && obj.repoOrg && obj.branch ? (
+          <ExternalLink
+            href={`https://github.com/${obj.repoOrg}/tree/${obj.branch}`}
+            text={`[${obj.branch}]`}
+          />
+        ) : (
```
Clicking on the branch value takes to 404 page not found and also this will work only for `github` provider, PAC supports gitlab, bitbucket urls as well.

abhinandan13jan commented:
<img width="1491" alt="Screenshot 2022-08-23 at 11 50 32 PM" src="https://user-images.githubusercontent.com/24852534/186450714-9fcaa5f9-37f2-496e-987a-a1b4b2a98a41.png">
<img width="1495" alt="Screenshot 2022-08-24 at 6 40 43 PM" src="https://user-images.githubusercontent.com/24852534/186450734-e8fe447e-a297-4078-814f-1af93c7249f2.png">
<img width="1498" alt="Screenshot 2022-08-24 at 6 40 54 PM" src="https://user-images.githubusercontent.com/24852534/186450742-23e03280-d89d-435c-b3ca-e6faf8910577.png">

christianvogt commented:
> 2. Clicking on `commitID` and `Component` links takes us to non-existing routes,   @christianvogt shall we link this later once we have the details page for  commit and components, wdyt?

Yes; remove the links for now until we have a valid page to go to.

karthikjeeyar commented on src/hacbs/components/Commits/CommitsListRow.tsx:
```diff
@@ -0,0 +1,73 @@
+    <>
+      <TableData className={commitsTableColumnClasses.name}>
+        <Link to={`/app-studio/commit?name=${obj.sha}`} title={obj.sha} className="pf-u-mr-sm">
+          {obj.sha.length > 14 ? `${obj.sha.slice(0, 7)}` : obj.sha}
```
sha is 40 characters long, so no need to have a check here.

```suggestion
          {`${obj.sha.slice(0, 7)}`}
```

abhinandan13jan replied:
updated

karthikjeeyar reviewed (commented):
Branch link is opening 404 not found page. check this  [comment](https://github.com/openshift/hac-dev/pull/174#discussion_r953437613)

abhinandan13jan commented:
@karthikjeeyar removed links for commit & component & fixed the branch link
<img width="1494" alt="Screenshot 2022-08-25 at 3 24 01 PM" src="https://user-images.githubusercontent.com/24852534/186636719-702610bb-63e2-40e0-b50f-dad9f8019e7b.png">
<img width="1495" alt="Screenshot 2022-08-25 at 3 24 12 PM" src="https://user-images.githubusercontent.com/24852534/186636728-eaabe24b-d032-453b-8eaf-f7974c0477e7.png">

abhinandan13jan commented:
@karthikjeeyar Updated PR, now looks like 
<img width="1490" alt="Screenshot 2022-08-25 at 7 35 33 PM" src="https://user-images.githubusercontent.com/24852534/186687218-329a20ef-a03e-4837-a473-1e01c2a56389.png">
<img width="1500" alt="Screenshot 2022-08-25 at 7 35 43 PM" src="https://user-images.githubusercontent.com/24852534/186687260-7788daa4-3e90-40a8-ac99-aca3b403206d.png">

abhinandan13jan commented on src/hacbs/components/Commits/__tests__/CommitsListView.spec.tsx:
```diff
@@ -0,0 +1,28 @@
+const commits = getCommitsFromPLRs(pipelineWithCommits);
+
+describe('CommitsListRow', () => {
+  it('renders correct commit data', () => {
```
Yes, we can add more tests with CommitsDetailsView when other missing Links are in place.

karthikjeeyar commented:
Additional tests to be handled in this ticket - https://issues.redhat.com/browse/HACBS-1052 

/lgtm

christianvogt commented:
/approve

abhinandan13jan commented:
/retest

abhinandan13jan commented:
E2e cypress
<img width="734" alt="Screenshot 2022-08-29 at 8 11 19 PM" src="https://user-images.githubusercontent.com/24852534/187227791-33a46cf3-84b6-4044-b370-e40f54838350.png">

christianvogt commented:
/lgtm
